File: cerberus-ui/core/src/specs/spec.ts
export interface ISpec<TContext> {
    isSatisfiedBy(context: TContext): boolean;
    and(spec: ISpec<TContext>): ISpec<TContext>;
    or(spec: ISpec<TContext>): ISpec<TContext>;
    not(): ISpec<TContext>;
}


export abstract class Spec<TContext> implements ISpec<TContext> {
    abstract isSatisfiedBy(context: TContext): boolean;
    and(spec: ISpec<TContext>): ISpec<TContext> {
       // return new AndSpec(this as unknown as ISpec<TContext>, spec);
        return new AndSpec([this as unknown as ISpec<TContext>, spec]);
    }
    or(spec: ISpec<TContext>): ISpec<TContext> {
        return new OrSpec([this as unknown as ISpec<TContext>, spec]);
    }
    not(): ISpec<TContext> {
        return new NotSpec(this as unknown as ISpec<TContext>);
    }
}
export class AndSpec<TContext> extends Spec<TContext> {
    constructor(
        private specs: ISpec<TContext>[],
    ) {
        super();
    }
    isSatisfiedBy(context: TContext): boolean {
        return this.specs.every(spec => spec.isSatisfiedBy(context));
    }
    and(spec: ISpec<TContext>): ISpec<TContext> {
        return new AndSpec([...this.specs, spec]);
    }
}

export class OrSpec<TContext> extends Spec<TContext> {
    constructor(
        private specs: ISpec<TContext>[],
    ) {
        super();
    }

    isSatisfiedBy(context: TContext): boolean {
        return this.specs.some(spec => spec.isSatisfiedBy(context));
    }
    or(spec: ISpec<TContext>): ISpec<TContext> {
        return new OrSpec([...this.specs, spec]);
    }
}

export class NotSpec<TContext> extends Spec<TContext> {
    constructor(private spec: ISpec<TContext>) {
        super();
    }

    isSatisfiedBy(context: TContext): boolean {
        return !this.spec.isSatisfiedBy(context);
    }
}

File: src/features/CerberusSurveillance/Features/Shared/Specs/Spec.cs
using System.Text.Json.Serialization;

namespace Cerberus.Surveillance.Features.Features.Shared.Specs;


public interface ISpec
{
    
}

/*[JsonPolymorphic(TypeDiscriminatorPropertyName = "__type")]
[JsonDerivedType(typeof(NotSpec<string>), "Not")]
[JsonDerivedType(typeof(AndSpec<string>), "And")]
[JsonDerivedType(typeof(OrSpec<string>), "Or")]
[JsonDerivedType(typeof(ValueEqualsSpec<string>), "Equals")]
[JsonDerivedType(typeof(ValueGraterThanSpec<string>), "GreaterThan")]
[JsonDerivedType(typeof(ValueLowerThanSpec<string>), "LowerThan")]*/
public abstract class Spec<T>: ISpec
{
    public abstract bool IsSatisfiedBy(T item);


    public static Spec<T>? operator &(Spec<T>? left, Spec<T>? right)
    {
        return left.And(right);
    }

    public static Spec<T>? operator |(Spec<T>? left, Spec<T>? right)
    {
        return left.Or(right);
    }

    public static Spec<T> operator !(Spec<T> spec)
    {
        return spec.Not();
    }
    
}